[docs] Uplift `TreeView` customization examples (#11424)
mui/mui-x: docs/data/tree-view/simple-tree-view/customization/customization.md

# Simple Tree View - Customization

> <p class="description">Learn how to customize the Tree View component.</p>

Reviewer: ```suggestion
<p class="description">Learn how to customize the simple version of the Tree View component.</p>
```

## Basics

### Custom icons

Use the `defaultCollapseIcon`, `defaultExpandIcon`, and `defaultEndIcon` props to customize the Tree View icons.

> You can use either icon libraries, such as [Material Icons](/material-ui/material-icons/), or create your own.

Reviewer: ```suggestion
The demo below shows how to add icons using both an existing icon library, such as [Material Icons](/material-ui/material-icons/), and creating an icon from scratch using Material UI's [SVG Icon component](/material-ui/icons/#svgicon).
```

{{"demo": "CustomIcons.js", "defaultCodeOpen": false}}

### Custom toggle animations

Use the `TransitionComponent` prop on the `TreeItem` to pass a component that handles your animation.

> The below demo animates Material UI's [Collapse](/material-ui/transitions/#collapse) with [react-spring](https://www.react-spring.dev/).

Reviewer: ```suggestion
The demo below is animated using Material UI's [Collapse](/material-ui/transitions/#collapse) component together with the [react-spring](https://www.react-spring.dev/) library.
```

{{"demo": "CustomAnimation.js", "defaultCodeOpen": false}}

### Custom styling

> Use the `treeItemClasses` to target internal elements of the Tree Item component and change their styles.

Reviewer: ```suggestion
Use `treeItemClasses` to target internal elements of the Tree Item component and change their styles.
```

Or we could alternatively say what `treeItemClasses` is, which I'm not sure whether it's a hook, class, etc.

```suggestion
Use the `treeItemClasses` {hook?}{class?} to target internal elements of the Tree Item component and change their styles.
```

Author: Technically it's an object that gives you the css classes that the `TreeItem` has, but I'm not sure it's worth specifying that, so I think we should go with the first suggestion 👍

{{"demo": "CustomStyling.js"}}

### Adding custom content

Use the `ContentComponent` prop and the `useTreeItem` hook to replace the Tree Item content with an entirely custom component.

> The demo below adds an avatar and custom typography.

Reviewer: ```suggestion
The demo below shows how to add an avatar and custom typography elements.
```

> Use the `ContentComponent` prop and the `useTreeItem` to replace the Tree Item content with an entirely custom component.

Reviewer: ```suggestion
Use the `ContentComponent` prop and the `useTreeItem` hook to replace the Tree Item content with an entirely custom component.
```

{{"demo": "CustomContentTreeView.js"}}

## Common examples

### Connection border

Target the `treeItemClasses.group` class to add connection borders between the Tree View items.

{{"demo": "BorderedTreeView.js", "defaultCodeOpen": false}}

### File explorer

The demo below shows many of the previous customization examples brought together to make the Tree View component look completely different than its default design.

{{"demo": "CustomizedTreeView.js"}}

### Gmail clone

Google's Gmail side nav is potentially one of the most famous tree view components.
The demo below shows how to replicate it.

{{"demo": "GmailTreeView.js"}}

> ### Limit expansion to icon container

Reviewer: How about moving this right below the "Connection border"? That way, we'd have somewhat of an order of complexity:

- Connection border: potentially the simplest of the common examples
- Then this one: we've already introduced the "ContentComponent" prop above, so I think it's okay to omit it in the description and focus on its difference.
- File explorer: several customizations brought together, more fleshed out
- Gmail clone: basically the same as above but replciating an existing design

> Google's Gmail side nav is potentially one of the most famous tree view components.

Reviewer: ```suggestion
Google's Gmail side nav is potentially one of the web's most famous tree view components.
```

> The demo below shows how to use a custom content for the Tree View, and limit the expansion of the Tree Items to the `iconContainer`

Reviewer: ```suggestion
The demo below shows how to trigger the expansion interaction just by clicking on the icon container instead of the whole Tree Item surface.
```

<p class="description">Customize the rendering of your Tree View.</p>

## Basics of customization

### Custom icons

> Use the `defaultCollapseIcon`, `defaultExpandIcon`, and `defaultEndIcon` props to customize the Tree View icons. You can use either icon libraries, such as Material Icons, or create your own.

Author: I was thinking about it too, but it's doable in multiple ways so in this case I feel like 'create your own' is enough 🤔

{{"demo": "CustomIcons.js", "defaultCodeOpen": false}}

### Custom toggle animations

Use the `TransitionComponent` prop on the `TreeItem` to pass a component that handles your animation.

{{"demo": "CustomAnimation.js", "defaultCodeOpen": false}}

> ### Styling

Reviewer: ```suggestion
### Custom styling
```

Maybe? Felt like we could continue the pattern here.

> You can use the `treeItemClasses` to target the internal elements of the TreeView.

Reviewer: ```suggestion
Use the `treeItemClasses` to target internal elements of the Tree Item component and change their styles.
```

PS: the demo looks a bit muddy in dark mode — maybe we should add the specific styles for it? It seems it's just light mode-focused.

<img width="250" alt="Screenshot 2024-01-12 at 08 02 20" src="https://github.com/mui/mui-x/assets/67129314/373a896d-628c-4118-b08c-e6a16f5cac7e">

{{"demo": "CustomStyling.js"}}

> ### Setting custom content

Reviewer: ```suggestion
### Adding custom content
```

> You can use the `ContentComponent` prop to replace the `TreeItem` with a custom component

Reviewer: ```suggestion
Use the `ContentComponent` prop to replace the Tree Item content with an entirely custom component.
The demo below adds an avatar and custom typography.
```

{{"demo": "CustomContentTreeView.js"}}

## Common examples

### Connection border

Target the `treeItemClasses.group` class to add connection borders between the Tree View items.

{{"demo": "BorderedTreeView.js", "defaultCodeOpen": false}}

### File explorer

> You can customize the Tree View to accomodate more complex use cases:

Reviewer: ```suggestion
The demo below shows many of the previous customization examples brought together to make the Tree View component look completely different than its default design.
```

{{"demo": "CustomizedTreeView.js"}}

> ### Gmail clone

Reviewer: ```suggestion
### Gmail clone

Google's Gmail side nav is potentially one of the most famous tree view components.
The demo below shows how to replicate it.
```

Feel like it's a bit weird not having at least a small description text there :)

{{"demo": "GmailTreeView.js"}}

> ### ContentComponent prop

Reviewer: Feeling like this being part of the Common examples is not fitting... also given we have already introduced this prop on the last demo of the Basics section. How about adding just one of the demo below, right after the "Adding custom content" demo, and dedicated to the `useTreeItem` hook?

Author: Hmm, I think the demo in the basics section already uses the `useTreeItem` hook. How about rephrasing the description there, and rebranding this one like Flavien suggested? (Did it in the latest commit)

Reviewer: I think that works! Throwing in an additional round of comments super quick :)
